src/components/KeywordHighlighterComponent.js

```diff
@@ -0,0 +1,255 @@
+ *  highlighting
+ * @property {bool} on flag which is going to be used to check whether to highlight
+ *  the appropriate sentences
+ */
+class KeywordHighlighterComponent {
+  constructor() {
+    this.keywords = [];
+    this.on = false;
+
+    this.#registerEvents();
+  }
+
+  /**
+   * Sets the keywords that the highlighting is going to use.
+   *
+   * @param {String[]} keywords
+   */
+  setKeywords(keywords) {
+    this.keywords = keywords;
+  }
+
+  /**
+   * Shows the highlited sentences, text layers processed next will have their
+   * highlited sentences turned on by default.
+   */
+  turnOn() {
+    this.on = true;
+    let highlightedTexts = document.querySelectorAll(
+      `.${TOPIC_HIGHLIGHTED_TEXT_CLASS}`
+    );
+    highlightedTexts.forEach((highlightedText) => {
+      highlightedText.classList.add(TOPIC_HIGHLIGHT_ACTIVE_CLASS);
+    });
+  }
+
+  /**
+   * Hides the highlited sentences, text layers processed next will have their
+   * highlited sentences turned off by default.
+   */
+  turnOff() {
+    this.on = false;
+    let highlightedTexts = document.querySelectorAll(
+      `.${TOPIC_HIGHLIGHTED_TEXT_CLASS}`
+    );
+    highlightedTexts.forEach((highlightedText) => {
+      highlightedText.classList.remove(TOPIC_HIGHLIGHT_ACTIVE_CLASS);
+    });
+  }
+
+  /**
+   * Registers for 'onTextLayerRendered' events to process the rendered text layer
+   * and inject HTML to highlight the sentences.
+   */
+  #registerEvents() {
+    EventHandlerService.subscribe(
+      PDFLEvents.onTextLayerRendered,
+      this.#onTextLayerRendered.bind(this)
+    );
+  }
+
+  /**
+   * Takes a text layer generated by pdf.js library, processes it by injecting
+   * HTML spans to sentences that need to be highlighted.
+   *
+   * Text layer consists of a list of spans called 'presentations'. Presentations
+   * are span elements containing text. There is no rule for how long the text is
+   * going to be in these presentations, or where the end of the sentence is.
+   *
+   * This function goes through all the child 'presentations' and if one contains
+   * a keyword, it will delegate work to other helper functions for injecting HTML
+   * span elements around text that needs to be highlighted.
+   *
+   * @param {HTMLElement} textLayer
+   */
+  #onTextLayerRendered(textLayer) {
+    for (let currentPresentation of textLayer.children) {
+      if (currentPresentation.tagName == "SPAN") {
+        let content = currentPresentation.innerHTML;
+        for (let keyword of this.keywords) {
+          let keywordIndex = content.indexOf(keyword);
+          if (keywordIndex != -1) {
+            let dotIndex = currentPresentation.innerHTML.indexOf(". ");
+            if (dotIndex == -1) {
+              this.#highlightUntilLeftDot(currentPresentation);
+              this.#highlightUntilRightDot(
+                currentPresentation.nextElementSibling
+              );
+            } else {
+              if (keywordIndex > dotIndex) {
```

Comment:
How about just else if and else instead of else block and then if and else in it..?

Reply:
I don't understand

Reply:
Your for loop could look like:
```
for (let keyword of this.keywords) {
   let keywordIndex = content.indexOf(keyword);
   if (keywordIndex != -1) {
      let dotIndex = currentPresentation.innerHTML.indexOf(". ");
      if (dotIndex == -1) {
         this.#highlightUntilLeftDot(currentPresentation);
         this.#highlightUntilRightDot(
             currentPresentation.nextElementSibling
           );
      } else if (keywordIndex > dotIndex) {
         this.#highlightToLeftDot(currentPresentation);
         this.#highlightUntilRightDot(
            currentPresentation.nextElementSibling
          );
       } else {
          this.#hightlightToRightDot(currentPresentation);
          this.#highlightUntilLeftDot(
		currentPresentation.previousElementSibling
	   );
	}
     }
  }
```

or even:
```
for (let keyword of this.keywords) {
   let keywordIndex = content.indexOf(keyword);
   if (keywordIndex == -1) continue;
   let dotIndex = currentPresentation.innerHTML.indexOf(". ");
   if (dotIndex == -1) {
      this.#highlightUntilLeftDot(currentPresentation);
      this.#highlightUntilRightDot(
          currentPresentation.nextElementSibling
        );
   } else if (keywordIndex > dotIndex) {
      this.#highlightToLeftDot(currentPresentation);
      this.#highlightUntilRightDot(
         currentPresentation.nextElementSibling
       );
   } else {
       this.#hightlightToRightDot(currentPresentation);
       this.#highlightUntilLeftDot(
	   currentPresentation.previousElementSibling
	);
   }
}
```

Point I'm trying to make it to lose unnecessary bloks/indentations since they usually lead to less readable code.